src/execution_plan/ops/op_apply.c

```diff
@@ -7,97 +7,91 @@
 #include "op_apply.h"
 
 /* Forward declarations. */
+static OpResult ApplyInit(OpBase *opBase);
 static Record ApplyConsume(OpBase *opBase);
 static OpResult ApplyReset(OpBase *opBase);
 static void ApplyFree(OpBase *opBase);
 
 OpBase *NewApplyOp(const ExecutionPlan *plan) {
 	Apply *op = rm_malloc(sizeof(Apply));
-	op->init = true;
-	op->rhs_idx = 0;
-	op->lhs_record = NULL;
-	op->rhs_records = array_new(Record, 32);
+	op->r = NULL;
+	op->op_arg = NULL;
+	op->bound_branch = NULL;
+	op->rhs_branch = NULL;
 
 	// Set our Op operations
-	OpBase_Init((OpBase *)op, OPType_APPLY, "Apply", NULL, ApplyConsume, ApplyReset, NULL, NULL,
+	OpBase_Init((OpBase *)op, OPType_APPLY, "Apply", ApplyInit, ApplyConsume, ApplyReset, NULL, NULL,
 				ApplyFree, false, plan);
 
 	return (OpBase *)op;
 }
 
-static Record ApplyConsume(OpBase *opBase) {
-	Apply *op = (Apply *)opBase;
-
-	assert(op->op.childCount == 2);
+static OpResult ApplyInit(OpBase *opBase) {
+	assert(opBase->childCount == 2);
 
-	Record rhs_record;
-
-	if(op->init) {
-		// On the first call to ApplyConsume, we'll read the entirety of the
-		// right-hand stream and buffer its outputs.
-		op->init = false;
+	Apply *op = (Apply *)opBase;
+	/* The op's bound branch and optional match branch have already been built as
+	 * the Apply op's first and second child ops, respectively. */
+	op->bound_branch = opBase->children[0];
+	op->rhs_branch = opBase->children[1];
 
-		OpBase *rhs = op->op.children[1];
-		while((rhs_record = rhs->consume(rhs))) {
-			op->rhs_records = array_append(op->rhs_records, rhs_record);
-		}
-	}
+	// Locate branch's Argument op tap.
+	op->op_arg = (Argument *)ExecutionPlan_LocateOp(op->rhs_branch, OPType_ARGUMENT);
+	assert(op->op_arg && op->op_arg->op.childCount == 0);
 
-	uint rhs_count = array_len(op->rhs_records);
+	return OP_OK;
+}
 
-	OpBase *lhs = op->op.children[0];
-	if(op->lhs_record == NULL) {
-		// No pending data from left-hand stream
-		op->rhs_idx = 0;
-		op->lhs_record = lhs->consume(lhs);
-	}
+static Record ApplyConsume(OpBase *opBase) {
+	Apply *op = (Apply *)opBase;
 
-	if(op->lhs_record == NULL) {
-		// Left-hand stream has been fully depleted
-		return NULL;
-	}
+	while(true) {
+		if(op->r == NULL) {
+			// Retrieve a Record from the bound branch if we're not currently holding one.
+			op->r = OpBase_Consume(op->bound_branch);
+			if(!op->r) return NULL; // Bound branch and this op are depleted.
 
-	// Clone the left-hand record
-	Record r = OpBase_CloneRecord(op->lhs_record);
+			// Successfully pulled a new Record, propagate to the top of the RHS branch.
+			if(op->op_arg) Argument_AddRecord(op->op_arg, OpBase_CloneRecord(op->r));
```

Comment:
`ApplyInit ` verified that `op->op_arg` isn't `NULL`
`assert(op->op_arg && op->op_arg->op.childCount == 0);`